enhancements/kube-apiserver/bound-sa-tokens.md

```diff
@@ -0,0 +1,146 @@
+---
+title: bound-service-account-tokens
+authors:
+  - "@marun"
+reviewers:
+  - "@deads2k"
+  - "@mfojtik"
+  - "@stlaz"
+  - "@sttts"
+approvers:
+  - TBD
+creation-date: 2019-11-28
+last-updated: 2019-11-28
+status: provisional
+see-also:
+  - "https://github.com/kubernetes/community/blob/master/contributors/design-proposals/auth/bound-service-account-tokens.md"
+  - "https://docs.google.com/document/d/1XcOsEv4jO9P1QQHn-tOnC80oMyCm85hGA6LqHRfjTgo/edit?ts=5ddb86c1"
+  - "https://aws.amazon.com/blogs/opensource/introducing-fine-grained-iam-roles-service-accounts/"
+  - "https://thenewstack.io/no-more-forever-tokens-changes-in-identity-management-for-kubernetes/"
+  - "https://jpweber.io/blog/a-look-at-tokenrequest-api/"
+replaces:
+superseded-by:
+---
+
+# Bound Service Account Tokens
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Open Questions
+
+- Is it necessary/desirable to allow `service-account-max-token-expiration` to be configured?
+
+## Summary
+
+Enable the optional use of bound service account tokens via volume projection and the
+TokenRequest API.
+
+## Motivation
+
+3rd party IAM components such as the [AWS pod identity
+webhook](https://github.com/aws/amazon-eks-pod-identity-webhook) require bound tokens to
+be able to identify pods without the use of workarounds such as proxies. It is also
+generally desirable to support bound tokens to limit the scope of permissions (and
+therefore risk of compromise) for a given service account token.
+
+### Goals
+
+- A pod can request a bound service account token via a projected volume.
+- A bound token can be acquired via the TokenRequest API.
+- The public key used to validate bound tokens can be retrieved for use with 3rd party IAM
+  components.
+
+### Non-Goals
+
+- Integrate or ship the aws pod identity webhook
+- Convert existing token-consuming components to use bound tokens
+
+## Proposal
+
+- Support configuring the options (via `KubeAPIServerConfig`) that will enable bound
```

Comment:
be clear, is this for a user to configure it or for the operator to wire it?

Reply:
I'm thinking the operator must configure the private key, and should set sane defaults for audiences and issuer.  I'm not sure of the value of user-provided audience/api, I'm curious if you think those should be user-configurable.

Reply:
I think we need a way to configure the issue URL see https://jira.coreos.com/browse/MSTR-893?focusedCommentId=152010&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-152010


I think the key, audiences can be defaulted and proabably don't even need to be configurable.

Reply:
As per my comment on your comment, I don't think it is necessary to configure the issuer URL for the intended use with the aws pod identity webhook. It will be the responsibility of the integrator to publish the verifying public key (e.g. via S3 as per the [webhook docs](https://github.com/aws/amazon-eks-pod-identity-webhook/blob/master/SELF_HOSTED_SETUP.md#public-issuer). Maybe it will make sense to make the issuer configurable in the future if openshift were to take responsibility for exposing the public key via an OIDC discovery endpoint, but @deads2k indicated that is out of scope for this enhancement.

Reply:
from my original comment in JIRA..

The `iss` in the tokens needs to match the URL provided to the IAM endpoint. So i do think there is a requirement that the issuer URL be set to the place where the public cert is hosted.

Reply:
My apologies, you are correct. I was assuming that the OIDC publishing documented by the webhook was AWS-specific, and didn't understand that however it would be done would be shared by all integrations. So yes, as you say, `--service-account-issuer` will need to be configurable.

Reply:
- If I were a customer with multiple clusters and wanted AWS IAM integration with all of them, would I want to use a different issuer for each of them or reuse the same one? 

- If reusing the same issuer, would I want to share the bound token keypair across clusters or supply the public keys of multiple keypairs via the OIDC endpoint?

Reply:
Would it make sense to disable bound tokens if `--service-account-issuer` is not set? Or should the feature be enabled by default with `iss` defaulted if not provided on initial deployment, and it could be changed from the default subsequently if required?

Reply:
> Would it make sense to disable bound tokens if `--service-account-issuer` is not set? Or should the feature be enabled by default with `iss` defaulted if not provided on initial deployment, and it could be changed from the default subsequently if required?

I would have it mostly configured with something like example.com so the URL dead-ends